inbm/dispatcher-agent/dispatcher/update_logger.py

```diff
@@ -105,44 +105,47 @@ def update_granular_with_sota_package_upgrade(self) -> None:
-
-                # Get status using dpkg-query
-                status = check_package_status(package_name)
-
-                package_info = {
-                    "update_type": OS,
-                    "package_name": package_name,
-                    "update_time": update_time,
-                    "action": PACKAGE_UPGRADE,
-                    "status": status,
-                    "version": version
-                }
-                # Remove previous entries with the same package_name from the UpdateLog list
-                data['UpdateLog'] = [log for log in data['UpdateLog'] if
-                                     log['package_name'] != package_info['package_name']]
-                # Append the new package information to the UpdateLog list
-                data['UpdateLog'].append(package_info)
-
-            # Open the file in write mode to save the updated data
-            with open(GRANULAR_LOG_FILE, 'w') as f:
-                json.dump(data, f, indent=4)
+        try:
+            if log_data:
+                # Get the latest upgrade history
+                # Split the log into parts
+                parts = log_data.split('\n\n')
+                # Filter parts that contain 'upgrade'
+                upgrade_parts = [part for part in parts if 'upgrade' in part.lower()]
+                latest_upgrade = upgrade_parts[-1]
+                update_time = get_package_start_date(latest_upgrade)
+                package_dict = extract_package_names_and_versions(latest_upgrade)
+
+                # Load current data in granular log file.
+                with open(GRANULAR_LOG_FILE, 'r') as f:
+                    data = json.load(f)
+
+                # Create the package information and store it
+                for package_name, version in package_dict.items():
+                    logger.debug(f"Package: {package_name}, Version: {version}")
+
+                    # Get status using dpkg-query
+                    status = check_package_status(package_name)
+
+                    package_info = {
+                        "update_type": OS,
+                        "package_name": package_name,
+                        "update_time": update_time,
+                        "action": PACKAGE_UPGRADE,
+                        "status": status,
+                        "version": version
+                    }
+                    # Remove previous entries with the same package_name from the UpdateLog list
+                    data['UpdateLog'] = [log for log in data['UpdateLog'] if
+                                         log['package_name'] != package_info['package_name']]
+                    # Append the new package information to the UpdateLog list
+                    data['UpdateLog'].append(package_info)
+
+                # Open the file in write mode to save the updated data
+                with open(GRANULAR_LOG_FILE, 'w') as f:
+                    json.dump(data, f, indent=4)
+        except IndexError as e:
```

Comment:
it is hard to figure out from reading this code where the IndexError is being thrown. could you instead check that the key exists inside the code explicitly?

Reply:
There are two places where an IndexError can occur. 

The first place is https://github.com/intel/intel-inb-manageability/blob/e430c83778890452a8b14b69518104ab156e42d8/inbm-lib/inbm_lib/package_info.py#L40

The second place is https://github.com/intel/intel-inb-manageability/blob/e430c83778890452a8b14b69518104ab156e42d8/inbm/dispatcher-agent/dispatcher/update_logger.py#L115

In both cases, the code is trying to retrieve a key from a list. 
I will add the key checking and raise a KeyError.